cmd/gpu-kubelet-plugin/vfio-device.go

```diff
@@ -0,0 +1,286 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"k8s.io/klog/v2"
+	cdiapi "tags.cncf.io/container-device-interface/pkg/cdi"
+	cdispec "tags.cncf.io/container-device-interface/specs-go"
+)
+
+const (
+	kernelIommuGroupPath   = "/sys/kernel/iommu_groups"
+	vfioPciModule          = "vfio_pci"
+	vfioPciDriver          = "vfio-pci"
+	nvidiaDriver           = "nvidia"
+	sysModulesRoot         = "/sys/module"
+	pciDevicesRoot         = "/sys/bus/pci/devices"
+	vfioDevicesRoot        = "/dev/vfio"
+	unbindFromDriverScript = "/usr/bin/unbind_from_driver.sh"
+	bindToDriverScript     = "/usr/bin/bind_to_driver.sh"
+	driverResetRetries     = "5"
+	gpuFreeCheckInterval   = 1 * time.Second
+	gpuFreeCheckTimeout    = 60 * time.Second
+)
+
+type VfioPciManager struct {
+	driverRoot string
+	driver     string
+	nvlib      *deviceLib
+}
+
+func NewVfioPciManager(driverRoot string, nvlib *deviceLib) *VfioPciManager {
+	vm := &VfioPciManager{
+		driverRoot: driverRoot,
+		driver:     vfioPciDriver,
+		nvlib:      nvlib,
+	}
+	if !vm.isVfioPCIModuleLoaded() {
+		err := vm.loadVfioPciModule()
+		if err != nil {
+			klog.Fatalf("failed to load vfio_pci module: %v", err)
+		}
+	}
+	return vm
+}
+
+// PreChecks tests if vfio-pci device allocations can be used.
+func (vm *VfioPciManager) Prechecks() error {
+	if !vm.isVfioPCIModuleLoaded() {
+		return fmt.Errorf("vfio_pci module is not loaded")
+	}
+	iommuEnabled, err := vm.isIommuEnabled()
+	if err != nil {
+		return err
+	}
+	if !iommuEnabled {
+		return fmt.Errorf("IOMMU is not enabled in the kernel")
+	}
+	return nil
+}
+
+func (vm *VfioPciManager) isIommuEnabled() (bool, error) {
+	f, err := os.Open(kernelIommuGroupPath)
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	defer f.Close()
+	_, err = f.Readdirnames(1)
+	if err == io.EOF {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (vm *VfioPciManager) isVfioPCIModuleLoaded() bool {
+	modules, err := os.ReadDir(sysModulesRoot)
+	if err != nil {
+		return false
+	}
+
+	for _, module := range modules {
+		if module.Name() == vfioPciModule {
+			return true
+		}
+	}
+
+	return false
+
+}
+
+func (vm *VfioPciManager) loadVfioPciModule() error {
+	_, err := execCommandWithChroot(vm.driverRoot, "modprobe", []string{vfioPciModule}) //nolint:gosec
```

Comment:
This assumption doesn’t hold if the driverRoot is not the host root (for example, when the NVIDIA driver is installed at a custom path such as `/home/kubernetes/bin/nvidia`, as on GKE). It’s better to mount the host root explicitly to avoid this issue.